velox/exec/tests/utils/LocalExchangeSource.cpp

```diff
@@ -182,40 +181,61 @@ class LocalExchangeSource : public exec::ExchangeSource {
     };
   }
 
-  /// Stops timeout thread and makes sure there are no references to
-  /// sources or their callbacks in global state.
+  // Invoked to start exchange source to run. It clears 'stop_' which allows
+  // the background timeout executor to run on the first exchange source
+  // request.
+  static void start() {
+    std::lock_guard<std::mutex> l(mutex_);
+    VELOX_CHECK_NULL(timeoutCheckExecutor_);
+    VELOX_CHECK(timeouts_.empty());
+    stop_ = false;
+  }
+
+  // Invoked to stop the exchange source. It sets 'stop_', clears the current
+  // pending request 'timeouts_' and join/destroy 'timeoutCheckExecutor_' if
+  // created.
   static void stop() {
-    if (executor_) {
+    {
+      std::lock_guard<std::mutex> l(mutex_);
+      if (stop_) {
+        return;
+      }
       stop_ = true;
-      executor_->join();
+      if (timeoutCheckExecutor_ == nullptr) {
+        VELOX_CHECK(timeouts_.empty());
+        return;
+      }
       timeouts_.clear();
-      executor_.reset();
     }
+    timeoutCheckExecutor_->join();
+    timeoutCheckExecutor_.reset();
   }
 
  private:
   using ResultCallback = std::function<void(
       std::vector<std::unique_ptr<folly::IOBuf>> data,
       int64_t sequence,
       std::vector<int64_t> remainingBytes)>;
+
   static void registerTimeout(
       const std::shared_ptr<ExchangeSource>& self,
       ResultCallback callback,
       std::chrono::microseconds maxWait) {
-    std::lock_guard<std::mutex> l(timeoutMutex_);
-    if (!executor_) {
-      executor_ = std::make_unique<folly::CPUThreadPoolExecutor>(1);
+    std::lock_guard<std::mutex> l(mutex_);
+    VELOX_CHECK(!stop_, "Local exchange source has stopped");
```

Comment:
I'm seeing some test failed due to this check. Could it be the testingStopLocalExchange be called in the wrong order?

Reply:
We need to patch this up if we see this check fails in the future.  Other than that it looks good.

Reply:
I have put the start/shutdown into OperatorTestBase test case start and shutdown.